server/main.py

```diff
@@ -1,98 +1,174 @@
+        raise BadInputException("Report must be a dict, and must contain 'tokens', 'type', and 'result'.")
+    tokens = report['tokens']
+    report_type = report['type']
+    result = report['result']
+
+    if not isinstance(tokens, list):
+        raise BadInputException("Tokens must be a list.")
+
+    for token in tokens:
+        if not isinstance(token, dict) or 'preimage' not in token or not isinstance(token['preimage'], str):
+            raise BadInputException("POSTed tokens must have a preimage.")
+
+        try:
+            b64decode(token['preimage'], validate=True)
+        except Exception as e:
+            raise BadInputException("Could not parse token preimage.")
+
+        if any(latlong in token and not isinstance(token[latlong], (float, int)) for latlong in ('lat', 'long')):
+            raise BadInputException("Could not parse token location.")
+
+        if not isinstance(report_type, str) or report_type not in types.keys():
+            raise BadInputException("Could not parse token type.")
+
+        if not isinstance(result, str) or result not in results.keys():
+            raise BadInputException("Could not parse token result.")
+
+    store_result(tokens, results[result], types[report_type])
+
 
 def tokens(request):
-  if request.method == 'GET':
-    with db_connect() as conn:
-      all_tokens = conn.execute("SELECT token_value, report_type FROM token WHERE report_result = 'POSITIVE'").fetchall()
-      return jsonify({'tokens': [{'value': b64encode(row[0]).decode('ascii'), 'type': row[1]} for row in all_tokens]})
-  elif request.method == 'POST':
-    # Validate input
-    try:
-      request_json = request.get_json()
-      if not request_json or not isinstance(request_json, dict) or 'report' not in request_json:
-        raise BadInputException("Request must contain JSON, and must have a report.")
-      report = request_json['report']
-      process_report(report)
-    except (BadInputException, BinasciiError, ValueError) as err:
-      return ('Bad Request: ' + str(err), 400)
-    
-    return 'Success!'
-  else:
-    return ('Method Not Allowed', 405)
+    if request.method == 'GET':
+        with session_scope() as session:
+            toks = [{
+              'value': b64encode(t.value).decode('ASCII'),
+              'type': t.type
+            } for t in session.query(Token).all()]
 
+            return json.dumps({'tokens': toks})
+    elif request.method == 'POST':
+        # Validate input
+        try:
+            request_json = request.get_json()
+            if not request_json or not isinstance(request_json, dict) or 'report' not in request_json:
+                raise BadInputException("Request must contain JSON, and must have a report.")
+            report = request_json['report']
+            process_report(report)
+        except (BadInputException, BinasciiError, ValueError) as err:
+            return ('Bad Request: ' + str(err), 400)

```

Comment:
might make sense to have all validation in a single place; now it is split between here and process_report?

Reply:
Consolidated to only have a `try-except` in the `tokens` func.